### PR TITLE
Metadata cascade policy

### PR DESCRIPTION
**Pull request by carycheng**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @carycheng, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: Box.V2/Managers/BoxMetadataCascadePolicyManager.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                return response.ResponseObject;
+            }
+        }
+
+        /// <summary>
+        /// If a policy already exists on the specified folder, this will apply that new policy to the folder and the sub-folder items. 
+        /// </summary>
+        /// <param name="policyId">The policy Id to force apply.</param>
+        /// <param name="conflictResolution">The desired behavior if a conflict exists. Set to either "none" or "overwrite".</param>
+        public async Task<bool> ForceApplyCascadePolicyAsync(string policyId, string conflictResolution)
```
Comment (mattwiller) [inline Box.V2/Managers/BoxMetadataCascadePolicyManager.cs]: Do we want to create an enum of possible resolution values?

File: Box.V2/Managers/BoxMetadataCascadePolicyManager.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        public async Task<BoxMetadataCascadePolicy> CreateCascadePolicyAsync(string folderId, string scope, string templateKey)
+        {
+            folderId.ThrowIfNullOrWhiteSpace("folderId");
+            scope.ThrowIfNullOrWhiteSpace("scope");
+            templateKey.ThrowIfNullOrWhiteSpace("templateKey");
+
+            dynamic jsonObject = new JObject();
+            jsonObject.folder_id = folderId;
+            jsonObject.scope = scope;
+            jsonObject.template_key = templateKey;
```
Comment (mattwiller) [inline Box.V2/Managers/BoxMetadataCascadePolicyManager.cs]: This is incorrect — the API endpoint uses the non-standard casing `templateKey` for this parameter

File: Box.V2.Test/BoxMetadataCascadePolicyManagerTest.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        [TestMethod]
+        [TestCategory("CI-UNIT-Test")]
+        public async Task DeleteMetadataCascadePolicy_ValidResponse()
+        {
+            string responseString = "";
+            IBoxRequest boxRequest = null;
+            Uri metadataCascadePolicyUri = new Uri(Constants.MetadataCascadePolicyEndpointString);
+            Config.SetupGet(x => x.MetadataCascadePolicyUri).Returns(metadataCascadePolicyUri);
```
[the remark this answers is not in the document]
Comment (carycheng) [reply Box.V2.Test/BoxMetadataCascadePolicyManagerTest.cs]: If taken out the test throws an exception - ArgumentNullException: Value cannot be null.

Parameter Name: baseUri

[the remark this answers is not in the document]
Comment (mattwiller) [reply Box.V2.Test/BoxMetadataCascadePolicyManagerTest.cs]: Does that indicate an issue with the Config or something?  Why would that error be thrown?

File: Box.V2/Config/Constants.cs [hunk 4]
```diff
@@ -880,5 +883,19 @@ public static class WhitelistDirections
             /// </summary>
             public const string Both = "both";
         }
+
+        /*** The desired conflict-resolution if a template already exists on a given file or folder.***/
+        public static class ConflictResolution
```
Comment (mattwiller) [inline Box.V2/Config/Constants.cs]: How would a user access this?  Can we make some of the tests/documentation reference this?

File: Box.V2/Managers/BoxMetadataCascadePolicyManager.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            policyId.ThrowIfNullOrWhiteSpace("policyId");
+            conflictResolution.ThrowIfNullOrWhiteSpace("conflictResolution");
+
+            dynamic jsonObject = new JObject();
+            jsonObject.conflict_resolution = conflictResolution;
+
+            string jsonString = jsonObject.ToString();
+
+            BoxRequest request = new BoxRequest(_config.MetadataCascadePolicyUri, policyId)
```
Comment (mattwiller) [inline Box.V2/Managers/BoxMetadataCascadePolicyManager.cs]: This doesn't look like the correct URL for this endpoint!  Please correct the endpoint here and manually test to ensure that this code works.

File: docs/metadata-cascade-policies.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
+Metadata Cascade Policies
+=========================
+
+A metadata cascade policy indicates if a metadata instance value should be cascaded to files and subfolders in a specific folder.
+
+Any user with edit permisions on a folder can create metadata cascade policies for that given folder. Policies are asigned to exactly one folder and exactly one metadata instance on that folder. It should be noted that there is some delay from file upload to metadata application.
```
Comment (mattwiller) [inline docs/metadata-cascade-policies.md]: [typo] "asigned" should be "assigned"

File: docs/metadata-cascade-policies.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+- [Get Metadata Cascade Policies for Folder](#get-metadata-cascade-policies-for-folder)
+- [Force Apply Metadata Cascade Policy](#force-apply-metadata-cascade-policy)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+Create Metadata Cascade Policy
+------------------------------
+
+To create a new metadata cascade policy, call
+`MetadataCascadePolicyManager.CreateCascadePolicyAsync(String folderId, string scope, string templateKey)`.
```
Comment (mattwiller) [inline docs/metadata-cascade-policies.md]: The types in the method signature should use consistent casing; I recommend sticking with "string" for all three

**Pull request merged**